adr/component_offerings.md

```diff
@@ -0,0 +1,134 @@
+# Offer both low-level and high-level components that are exported from each package.
+
+## Context
+
+The primary issue we are facing right now is determining how "high-level" the component should be. For example, should we create a single component with many props, or should we create many components to be composed together.
+
+This boils down to two main goals:
+
+-   Provide enough flexibility and low-level components for teams to quickly build and deliver applications without being blocked by us
+-   Provide quality high level components that allow teams to quickly build applications with the best possible interactions (UX) that are consistent, performant, accessible and extensible without the need of repeating themselves over and over again.
+
+```jsx
+<ClayCard>
+	<ClayCard.Title onClick={handleTitleClick}>
+		<ClayIcon symbol="check" />
+
+		{'Hello World'}
+	</ClayCard.Title>
+
+	<ClayCard.Body onClick={handleBodyClick}>
+		<ClayCard.Description>
+			{'Some cool descriptors'}
+		</ClayCard.Description>
+	</ClayCard.Body>
+</ClayCard>
+
+// vs
+<ClayCard
+	iconTitleLeftSymbol="check"
+	onTitleClick={handleTitleClick}
+	onBodyClick={handleBodyClick}
+	description="foo bar baz"
+	title="hello world"
+/>
+```
+
+## Decision
+
+The decision going forward is to create 2 sets of components
+
+-   low-level composable components such as `<ClayCard.Header>`, `<ClayCard.Footer>`
+-   simple high-level components that cover basic lexicon uses
+
+### What does this look like?
+
+#### Composable:
+
+![Composable Drop Down](./images/component_offerings_2.png)
+
+```jsx
+<ClayDropDown>
+	<ClayDropDown.Search
+		onChange={() => {}}
+		placeholder="Search"
+		spritemap={spritemap}
+		value={''}
+	/>
+
+	<ClayDropDown.ItemList>
+		{items.map((item, i) => (
+			<ClayDropDown.Item href={item.href} key={i}>
+				{item.label}
+			</ClayDropDown.Item>
+		))}
+	</ClayDropDown.ItemList>
+</ClayDropDown>
+```
+
+Composing gives the user more control over the layout and interactions.
+
+#### High-level:
+
+![High Level Drop Down](./images/component_offerings_1.png)
+
+```jsx
+<ClayDropDownWithPagination
+	delta={delta}
+	header="Favorites"
+	items={items} // array of objects with `label` and `href`
+	message={'Showing {0} of {1} Structures'}
+	title="Folder"
+/>
+
+// Compare if this component was composed of low-level components
+<ClayDropDown>
+	<ClayDropDown.Header header="Folder" />
+
+	<ClayDropDown.ItemList>
+		<ClayDropDown.Group header="Favorites">
+			{items.slice(0, page * delta).map((item, i) => (
+				<ClayDropDown.Item href={item.href} key={i}>
+					{item.label}
+				</ClayDropDown.Item>
+			))}
+		</ClayDropDown.Group>
+	</ClayDropDown.ItemList>
+
+	<ClayDropDown.Caption>
+		{`Showing ${page * delta} of ${items.length} structures`}
+	</ClayDropDown.Caption>
+
+	<ClayDropDown.Action onClick={() => setPage(page + 1)}>
+		{'Show More'}
+	</ClayDropDown.Action>
+</ClayDropDown>
+```
```

Comment:
There are some other patterns that we could consider (or at least list) too. Given some low-level components `<Foo.A>`, `<Foo.B>` etc, depending on how many high-level embodiments we wish to offer, we might want them to be:

- `<FooVariantA />`, `FooVariantB />` etc: ie. the option you've laid out above.
- `<Foo a={...} b={...} />`: a single high-level component that has a bunch of options (I think this is your `ClayCard` example higher up).
- `<Foo mode={....} />`: some "mode" or "type" enum that sets the high-level behavior (I don't think this one has been mentioned so far).

And there are ultimately so many ways to extend and compose in React (whether they be higher-order components, render props, function children, context wrappers etc), that an analysis of individual cases might suggest a bunch of alternatives.

Reply:
As far as I'm concerned, I don't know which pattern is the best but I'd rather focus on small components that can be composed; at least that's what I've been used to do with React.

That said; we should take into account the fact that the teams using these components have been used to Clay's "high level" components; how are we going to "sell" them the fact that they are going to have to write much more code than before to get basically the same result? (Even if they asked for React in the 1st place).

Reply:
@matuzalemsteles thanks for the comment.

 > I think our "sell" here for the low-level components are for just cases where the team needs to create other behaviors or customize their components in more detail.

From what I've experienced that has happened with most of the components for most of the teams. (I don't have exact numbers but we could actually look at the closed issues and PRs).

Reply:
@wincent 
> And there are ultimately so many ways to extend and compose in React (whether they be higher-order components, render props, function children, context wrappers etc), that an analysis of individual cases might suggest a bunch of alternatives.

You're totally right in that there are tons of different ways to accomplish this. From my preliminary brainstorming, I think the first option is the clearest straight forward path and most descriptive for users. I don't really like the "many options" or "mode" or "type" methods due to the amount of branching that ends up happening in the components. For example, if "searchable" is an option, we now have a large chunk of conditional code in a single component. Personally, I prefer having really descriptive components, even if that means we have to reduplicate some code. I find it a lot easier to maintain components that strive to implement a single variant.

@matuzalemsteles 
> it is still unclear in this ADR how the high-level components will be, we will deliver only a single high-level component following the Lexicon standards with their interactions

I would avoid aiming for 'only a single high-level component'. This is going to be harder to decide though and I don't think there is necessarily a clear path on how to decide where we break a component at. I think it might be best to aim each high-level component at a single variation, rather than trying to have multiple functionalities in the same component. For example, `ClayDropDown` might have `ClayDropDownBasic`, `ClayDropDownWithForm`, and `ClayDropDownWithPagination`. There may be some overlap, but I think we can aim each high-level component at a specific use case.

That make more sense?

Reply:
@julien I agree in that we should focus on maintaining the building blocks, however, we also need the high-level components because we run the risk of many implementations of the same basic functionality. Now if we can also try and make out high-level components simple and straight forward, we may have more components to maintain, but they should also be easier to maintain since we have the core building blocks.

Reply:
@bryceosterhaus that sounds good to me then.

Reply:
I think it makes sense to deliver components that deliver specific cases, but I'm not sure we should call them High-level. That's a bit my vision now.

* Intermediate layer are simple components that deliver only specific use cases. Composed of low-level components.
* High-level components are composed of low-level and intermediate layer.

Even if we do not get to deliver high-level, I think we should improve our speech in this sense, we have this intermediate layer following specific cases of the Lexicon component... For example the image below, when I see something high-level I consider the whole component `ClayDropdown` delivering everything the Lexicon expects of that component. They would be our v2 components for example but maybe there will be a higher or no cost to maintain.

![Layer Components@2x](https://user-images.githubusercontent.com/13750819/58329487-07ac8e80-7e0b-11e9-971c-506e39bdc216.png)

Reply:
@matuzalemsteles yeah I think its totally fine to name it something else, it was just the clearest way to me at the time to differentiate between the two types of components we will build. As far as us communicating this to users, I'm not sure we even need to go into this type of depth with them either. They will just want to know how to use a certain lexicon use case as quickly as possible. The low-level stuff really only gets brought up when they want a more custom implementation.

The main point of this ADR is to provide a reason why developers need to create two sets of components when working on a package, the `<ClayDropDown.Menu>` type components and the `<DropDownWithPagination>` type. Not sure we need to be too concerned with users reading our ADRs.

Reply:
I agree with @bryceosterhaus here. We don't even to give them names or market them any differently.

I think the consensus and the conclussion here is that we want to follow [atomic design](http://atomicdesign.bradfrost.com/chapter-2/) principles (which is what inspires Lexicon too) and:
- Start off very small composable atoms
- Compose them as necessary to offer molecules, organisms and templates as deemed fit to help developers implement Lexicon patterns.

@bryceosterhaus, what do you think?

Reply:
I do not mean that users need to read our ADRs, I really did not want to encourage it (but whoever wants to go deeper into Clay is a starting point) when I say we should improve our discourse, it will be how we will present the two levels of components for people, for being something unusual perhaps for some people, especially the intermediary layer.

Since we will no longer have a high-level component (a complete component with all Lexicon definitions) how will we address Taglibs? or is it to address here in that ADR?

Reply:
Hey @matuzalemsteles, we need to have and provide all levels that make sense to us. That includes what you call high-level components, but we can simply call them components. I don't see the need to differentiate them. 

We just need to add the proper documentation and examples for all of our components and how to leverage them. No need to expose this anymore than that.

Reply:
hey @jbalsas about following based on `atomic design`, I think that in a way we ended up following these principles making the components available... but I'm not sure if it would apply inside a component, maybe yes. I can not think well of so many variations (molecules, organisms and templates) so within the component...

When I talk about high-level and low-level is something common that some lib deliver, for example Tensorflow.js delivers its low-level layer with [Tensorflow Core](https://github.com/tensorflow/tfjs- core) and its high-level layer with [Tensorflow Layer](https://github.com/tensorflow/tfjs-layers). Enabling what we were initially talking about flexibility for low-level layers and fast delivery to the high-level layers.

I'm just calling it that because it's easier to identify the components 😅 and understand what they actually do. I understand that we should only make them available as components, but the point is that we have to explain why we are doing this and when we explain we will have to separate things to create clarity for people.